src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+
+  let quad (g1 : 'a t) (g2 : 'b t) (g3 : 'c t) (g4 : 'd t) : ('a * 'b * 'c * 'd) t =
+    (g1 >|= (fun a b c d -> (a, b, c, d))) <*> g2 <*> g3 <*> g4
+
+  let char : char t = int_range ~origin:(int_of_char 'a') 0 255 >|= char_of_int
+
+  (** The first characters are the usual lower case alphabetical letters to help shrinking. *)
+  let printable_chars =
+    (* Left and right inclusive *)
+    let range min max = List.init (max - min) (fun i -> char_of_int (i + min)) in
+    let a = 97 in
+    let z = 122 in
+    let lower_alphabet = range a z in
+    (* ' ' *)
+    let first_printable_char = 32 in
+    let before_lower_alphabet = range first_printable_char (a - 1) in
+    (* '~' *)
+    let last_printable_char = 126 in
+    let after_lower_alphabet = range (z + 1) last_printable_char in
+    let newline = ['\n'] in
+    (* Put alphabet first for shrinking *)
+    List.flatten [lower_alphabet; before_lower_alphabet; after_lower_alphabet; newline]
+
+  let printable : char t =
+    int_range ~origin:0 0 (List.length printable_chars - 1)
+    >|= List.nth printable_chars
+
+  let numeral : char t =
+    let zero = 48 in
+    let nine = 57 in
+    int_range ~origin:zero zero nine >|= char_of_int
+
+  let string_size ?(gen = char) (size : int t) : string t =
+    list_size size gen >|= (fun l -> List.to_seq l |> String.of_seq)
+
+  let string ?(gen : char t option) : string t = string_size ?gen nat
+
+  let string_of gen = string_size ~gen nat
+
+  let string_readable = string_size ~gen:char nat
+
+  let small_string ?gen st = string_size ?gen small_nat st
+
+  let small_list gen = list_size small_nat gen
+
+  let small_array gen = array_size small_nat gen
+
+  let join (gen : 'a t t) : 'a t =
+    gen >>= Fun.id
+
+  (* corner cases *)
+
+  let graft_corners (gen : 'a t) (corners : 'a list) () : 'a t =
+    let cors = ref corners in fun st ->
+      match !cors with [] -> gen st
+                     | e::l -> cors := l; Tree.pure e
+
+  let int_pos_corners = [0;1;2;max_int]
+
+  let int_corners = int_pos_corners @ [min_int]
+
+  let nng_corners () : int t = graft_corners nat int_pos_corners ()
+
+  (* sized, fix *)
+
+  let sized_size (size : int t) (gen : 'a sized) : 'a t =
+    size >>= gen
+
+  let sized (gen : 'a sized) : 'a t = sized_size nat gen
+
+  let fix f =
+    let rec f' n st = f f' n st in
+    f'
+
+  let generate ?(rand=Random.State.make_self_init()) ~(n : int) (gen : 'a t) : 'a list =
+    list_repeat n gen rand |> Tree.root
+
+  let generate1 ?(rand=Random.State.make_self_init()) (gen : 'a t) : 'a =
+    gen rand |> Tree.root
+
+  let generate_tree ?(rand=Random.State.make_self_init()) (gen : 'a t) : 'a Tree.t =
+    gen rand
+
+  let delay (f : unit -> 'a t) : 'a t = fun st -> f () st
+
+  let add_shrink_invariant (p : 'a -> bool) (gen : 'a t) : 'a t =
+    fun st -> gen st |> Tree.add_shrink_invariant p
+
+  include Qcheck_ops.Make(struct
```

Comment:
When you say "Get rid of that", you mean "always expose it", or "remove the feature altogether"?

Reply:
I mean we can directly write the let operators without the codegen magic of `Qcheck_ops` :slightly_smiling_face: . This would become `let (let * ) = …   let (let+) = …` rather than a functor inclusion.

Reply:
Done in eb2b14a641b5af868f57b59baa35724f1c917fee